### ISSUE TITLE
Any plans for wasm64 support?

### ISSUE THREAD

**Issue by tudor** [WebAssembly/wasi-sdk#185]
Are there any plans to support 64-bit builds (where memory indices are 64-bit, which means pointers are 64-bit instead of 32, and then you might as well make `long` 64-bit as well and go full `LP64`)?

Thanks!

**Comment by sunfishcode** [comment]
Yes; the main work right now is happening in upstream LLVM (eg. [here](https://reviews.llvm.org/rG670944fb20b226fc22fa993ab521125f9adbd30a)).

There will also need to be some amount of work to port wasi-libc. All the code in wasi-libc is written with a wasm64 target in mind. We'll need a 64-bit version of the WASI APIs as well, and the code for that is also all written with a wasm64 target in mind. So in theory this should be a straightforward process.

**Comment by tudor** [comment]
Exciting!

I have a strong vested interest in seeing WASM succeed as a server-side plugin (and write-once, run-anywhere code distribution) platform. There are a few missing pieces:

- 64-bit support
- threads
- sockets
- stack unwinding / exception handling

but it seems that all 4 areas are seeing a lot of active work, which means 2022 will be an exciting year for WASM.

**Comment by TimWhiting** [comment]
It's been a few years and the clang link seems to have been resolved. Any updates?

**Comment by abrown** [comment]
I don't see anyone working on this too closely at the moment: are you interested in adding support for `wasm64-wasi`?

**Comment by TimWhiting** [comment]
If it isn't a bunch of work. Are there assumptions being made that make this a big effort, or is it a simple change?

**Comment by mh4ck-Thales** [comment]
I'd be interested in seeing wasi-sdk support wasm64, if everything is ready for it. Seems that the first thing to do would be to make wasi-libc support wasm64 ; should we open an issue there ?

**Comment by abrown** [comment]
There's been some work already on that in https://github.com/WebAssembly/wasi-libc/pull/444 but there are concerns about how that is implemented so I think some discussion is needed. @loganek is also interested in making this happen and has added an agenda item to the WASI subgroup meeting to have that discussion: https://github.com/WebAssembly/meetings/pull/1479. Can you join that?

**Comment by lastmjs** [comment]
Status update? Is there an equivalent to `wasm32-wasip1` for Wasm64?

**Comment by alexcrichton** [comment]
Personally I would say that what you describe is inaccurate, and I can try to clarify why. Defining `wasm64-wasip1` requires a specification of what that even is, and the problem here is generally not as simple as "just go make those pointers that were 4 bytes 8-bytes now instead". Making the target workable also requires cross-ecosystem collaboration with embedders, toolchains, developers, libraries, etc. This is a large, concerted, effort to a goal that currently has no definition and also has no standards-based future. To the best of my knowledge the WASI subgroup has had no work, development, or discussions about furthering WASIp1 development in many years.

While, yes, this reaches the same conclusion that `wasm64-wasip1` isn't and won't be a target, the rationale here is different in significant ways. This is not an attempt to avoid keeping WASIp1 alive, it's a reflection of how it simply is not alive at the standards level. This is not an attempt to push users to WASIp2 or components, it's a reflection of how all development in the WASI subgroup is focused on components. It's not some "mysterious force" trying to hold things back, it's a reflection of the priorities of all the developers currently paid to work in this space (such as myself).

Users are always empowered to create forks to their own desires, as is the nature of open source software. I would personally want to clarify the consequences of doing so in an environment such as this:

* Forked targets are unlikely to see language integration. For example there likely won't be official targets or builds for Rust, Go, Python, C, C++, etc. Users will have to maintain custom toolchains for all languages they wish to support, and the difficulty of doing so can vary greatly from toolchain to toolchain.
* Runtime support is unlikely to be integrated. For example as a maintainer of Wasmtime I would not want to add support for 64-bit WASIp1 unless the WASI subgroup officially decided to do so. This means that such hypothetical integration would live in a third-party location, which users would have to integrate. Users would need a separate CLI, for example.
* Library suport is likely to languish behind. Without any official language integration libraries may be hesitant to add support for a target that has no official definition or means by which it can be tested. 

Naturally these consequences affect various use cases in different ways. If you're building a polyglot plugin system, it's probably a non-starter. If you only need a C toolchain, have zero dependencies, and only work with one runtime, you're probably much less affected. This means that a pragmatic choice here is situation-specific and subjective, there's no right answer.

----

What I'll reiterate again from my [previous comment](https://github.com/WebAssembly/wasi-sdk/issues/212#issuecomment-3900352021) that it's not like the path to `wasm64-wasip2` is so nebulous and murky it's untenable. I've described the initial spec work [here](https://github.com/WebAssembly/component-model/issues/22#issuecomment-3627314893). In my initial implementation of the component-model in Wasmtime I also added internal support/refactoring to support memory64 almost everywhere, and https://github.com/bytecodealliance/wasmtime/issues/4311 outlines the remaining work for Wasmtime itself to support this.

Put another way, `wasm64-wasip2` does not exist because no one has spent time and resources to invest in it. The development required to support it is predicted to be straightforward enough for a motivated individual to get started and see what snags, if any, arise. All of WASI is developed in the open in such a way that motivated parties may contribute to its development and help further the ecosystem. As with all other open source software, however, if users aren't willing to contribute to development themselves then they need to either (a) accept that features may take time or (b) work to convince existing developers to prioritize a particular feature over others.
